walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,169 @@
+  * or are longer than the requested meeting duration.
+  * First extracts the existing {@code TimeRange} that correspond to {@code Event} having
+  * overlapping attendees with the {@code request}; we only need to avoid conflicting with
+  * those {@code TimeRange} that involve attendees contained in the {@code request}.
+  * Then finds the available {@code TimeRange} sandwiched between an existing, ending
+  * {@code TimeRange} and an existing starting {@code TimeRange}.
+  *
+  * @param events Existing events that each occupy some {@code TimeRange}.
+  * @param request A meeting request for some mandatory and optional attendees and a duration.
+  * @return All available {@code TimeRange} that satisfy the meeting request.
+  */
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    Set<String> attendees = new HashSet<>(request.getAttendees());
+    long requestedDuration = request.getDuration();
+
+    if (Long.compare(requestedDuration, TimeRange.WHOLE_DAY.duration()) > 0) {
+      return Arrays.asList();
+    } else if (events.isEmpty() || attendees.isEmpty()) {
+      return Arrays.asList(TimeRange.WHOLE_DAY);
+    }
+
+    Collection<TimeRange> occupiedTimeRanges =
+        getRelevantTimeRangesFromEvents(events, attendees);
+
+    return getAvailabeTimeRanges(occupiedTimeRanges, requestedDuration);
+  }
+
+  /**
+  * Finds the available {@code TimeRange} having at least a length of {@code requestedDuration},
+  * given some already occupied {@code TimeRange}. The resultant list of available
+  * {@code TimeRange} is sorted in ascending chronological order.
+  */
+  private List<TimeRange> getAvailabeTimeRanges(Collection<TimeRange> occupiedTimeRanges,
+      long requestedDuration) {
+    if (occupiedTimeRanges.isEmpty()) {
+      return Arrays.asList(TimeRange.WHOLE_DAY);
+    }
+
+    List<TimeRange> availableTimeRanges = new LinkedList<>();
+    List<TimeRange> startOrderedTimeRanges = getOrderedTimeRanges(occupiedTimeRanges,
+                                                                  TimeRange.ORDER_BY_START);
+    List<TimeRange> endOrderedTimeRanges = getOrderedTimeRanges(occupiedTimeRanges,
+                                                                TimeRange.ORDER_BY_END);
+
+    // Potentially add the {@code TimeRange} before the first existing {@code TimeRange}.
+    TimeRange firstStartingTimeRange = startOrderedTimeRanges.get(0);
+    checkDurationAndAddAvailableTimeRange(requestedDuration,
+                                          TimeRange.START_OF_DAY, firstStartingTimeRange.start(),
+                                          availableTimeRanges);
+
+    // Find available {@code TimeRange} by looking at the available time between one ending
+    // {@code TimeRange} and the immediately following starting {@code TimeRange}.
```

Comment:
Reword this comment. The " and the immediately following starting {@code TimeRange}" doesn't really make sense.